shotover-proxy/src/transforms/redis/sink_single.rs

```
@@ -73,43 +87,235 @@ impl Transform for RedisSinkSingle {
             } else {
                 Box::pin(tcp_stream) as Pin<Box<dyn AsyncStream + Send + Sync>>
             };
-            self.outbound = Some(Framed::new(generic_stream, RedisCodec::new()));
+
+            let (outbound_tx, outbound_rx) = Framed::new(generic_stream, RedisCodec::new()).split();
+            let (response_messages_tx, response_messages_rx) = mpsc::unbounded_channel();
+            let (sent_message_type_tx, sent_message_type_rx) = mpsc::unbounded_channel();
+
+            tokio::spawn(
+                server_response_processing_task(
+                    outbound_rx,
+                    self.pushed_messages_tx.clone(),
+                    response_messages_tx,
+                    sent_message_type_rx,
+                )
+                .in_current_span(),
+            );
+            self.connection = Some(Connection {
+                response_messages_rx,
+                sent_message_type_tx,
+                outbound_tx,
+            })
+        }
+
+        let connection = self.connection.as_mut().unwrap();
+
+        for message in &mut message_wrapper.messages {
+            let ty = if let Some(Frame::Redis(RedisFrame::Array(array))) = message.frame() {
+                if let Some(RedisFrame::BulkString(bytes)) = array.first() {
+                    match bytes.to_ascii_uppercase().as_slice() {
+                        b"SUBSCRIBE" | b"PSUBSCRIBE" | b"SSUBSCRIBE" => MessageType::Subscribe,
+                        b"UNSUBSCRIBE" | b"PUNSUBSCRIBE" | b"SUNSUBSCRIBE" => {
+                            MessageType::Unsubscribe
+                        }
+                        b"RESET" => MessageType::Reset,
+                        _ => MessageType::Other,
+                    }
+                } else {
+                    MessageType::Other
+                }
+            } else {
+                MessageType::Other
+            };
+            connection.sent_message_type_tx.send(ty).unwrap();
```

It's possible to avoid these unwraps. Just wondering your reasoning for why it's not worth doing so?

---

yep, good point, if the task panics causing the send to fail then we can just return an error terminating the connection without bringing down all of shotover.